lib/promise/group.rb

```diff
@@ -21,31 +23,33 @@ def wait
       end
     end
 
-    private
-
-    def chain_inputs
-      on_fulfill = method(:on_fulfill)
-      on_reject = promise.public_method(:reject)
-      each_promise do |input_promise|
-        input_promise.then(on_fulfill, on_reject)
-      end
-    end
-
-    def on_fulfill(_result)
+    def promise_fulfilled(_value, _arg)
       @remaining -= 1
       if @remaining.zero?
         result = @inputs.map { |obj| promise?(obj) ? obj.value : obj }
         promise.fulfill(result)
       end
     end
 
+    def promise_rejected(reason, _arg)
+      promise.reject(reason)
+    end
+
+    private
+
+    def chain_inputs
+      each_promise do |input_promise|
+        input_promise.subscribe(self, nil, nil) if input_promise.pending?
+      end
+    end
+
     def promise?(obj)
       obj.is_a?(Promise)
     end
 
     def count_promises
       count = 0
-      each_promise { count += 1 }
+      each_promise { |input_promise| count += 1 if input_promise.pending? }
```

Comment:
The initializer can no longer just do `promise.fulfill(inputs)` `if @remaining.zero?`, since `inputs` may now contain non-pending promises.  Also, if any of those non-pending promises were rejected, then we would need to reject the result promise instead of fulfill it.

Reply:
👍 Looks like existing tests don't cover this. Let me quickly fix that up and add some test coverage here! ❤️